static/app/components/carousel.tsx

```diff
@@ -1,92 +1,115 @@
 import {IconArrow} from 'sentry/icons';
-import {ArrowProps} from 'sentry/icons/iconArrow';
 import {t} from 'sentry/locale';
 import {space} from 'sentry/styles/space';
 
-type Props = {
+interface CarouselProps {
   children?: React.ReactNode;
-};
+  /**
+   * This number determines what percentage of an element must be within the
+   * visible scroll region for it to be considered 'visible'. If it is visible
+   * but slightly off screen it will be skipped when scrolling
+   *
+   * For example, if set to 0.8, and 10% of the element is out of the scroll
+   * area to the right, pressing the right arrow will skip over scrolling to
+   * this element, and will scroll to the next invisible one.
+   *
+   * @default 0.8
+   */
+  visibleRatio?: number;
+}
 
-function Carousel({children}: Props) {
+function Carousel({children, visibleRatio = 0.8}: CarouselProps) {
   const ref = useRef<HTMLDivElement | null>(null);
-  const [anchorRefs, setAnchorRefs] = useState<HTMLElement[]>([]);
-  const [childrenRefs, setChildrenRefs] = useState<HTMLElement[]>([]);
-  const [isAtStart, setIsAtStart] = useState(true);
-  const [isAtEnd, setIsAtEnd] = useState(false);
 
+  // The visibility match up to the elements list. Visibility of elements is
+  // true if visible in the scroll container, false if outside.
+  const [childrenEls, setChildrenEls] = useState<HTMLElement[]>([]);
+  const [visibility, setVisibility] = useState<boolean[]>([]);
+
+  const isAtStart = visibility[0];
+  const isAtEnd = visibility[visibility.length - 1];
+
+  // Update list of children element
+  useEffect(
+    () => setChildrenEls(Array.from(ref.current?.children ?? []) as HTMLElement[]),
+    [children]
+  );
+
+  // Update the threshold list. This
   useEffect(() => {
     if (!ref.current) {
       return () => {};
     }
 
     const observer = new IntersectionObserver(
-      entries => {
-        entries.forEach(e => {
-          if (e.target.id === anchorRefs[0].id) {
-            setIsAtStart(e.isIntersecting);
-          } else if (e.target.id === anchorRefs[1].id) {
-            setIsAtEnd(e.isIntersecting);
-          }
-        });
-      },
-      {root: ref.current, threshold: [1]}
+      entries =>
+        setVisibility(currentVisibility =>
+          // Compute visibility list of the elements
+          childrenEls.map((child, idx) => {
+            const entry = entries.find(e => e.target === child);
+
+            // NOTE: When the intersection observer fires, only elements that
+            // have passed a threshold will be included in the entries list.
+            // This is why we fallback to the currentThreshold value if there
+            // was no entry for the child.
+            return entry !== undefined
+              ? entry.intersectionRatio > visibleRatio
+              : currentVisibility[idx] ?? false;
+          })
+        ),
+      {
+        root: ref.current,
+        threshold: [0, visibleRatio, 1],
```

Comment:
hmm why do we need the 0 & 1 thresholds? looks like we only need to update visibility when the interaction ratio crosses `visibleRatio`, so shouldn't `[visibleRatio]` suffice?

Reply:
oh good point 👍

Reply:
Fixed